images/src/org/intellij/images/util/imageio/svg/SvgImageReader.kt

```diff
@@ -0,0 +1,144 @@
+// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
+package org.intellij.images.util.imageio.svg
+
+import com.github.weisj.jsvg.attributes.font.SVGFont
+import com.github.weisj.jsvg.geometry.size.Length
+import com.github.weisj.jsvg.nodes.SVG
+import com.intellij.ui.paint.PaintUtil
+import com.intellij.ui.scale.ScaleContext
+import com.intellij.ui.svg.createJSvgDocument
+import com.intellij.util.ui.ImageUtil
+import java.awt.image.BufferedImage
+import java.io.IOException
+import java.io.InputStream
+import javax.imageio.ImageReadParam
+import javax.imageio.ImageReader
+import javax.imageio.ImageTypeSpecifier
+import javax.imageio.stream.ImageInputStream
+import kotlin.math.max
+
+class SvgImageReader(svgImageReaderSpi: SvgImageReaderSpi) : ImageReader(svgImageReaderSpi) {
+  private var height: Length? = null
+  private var width: Length? = null
+  private var jSvgDocument: SVG? = null
+  override fun setInput(input: Any?, seekForwardOnly: Boolean, ignoreMetadata: Boolean) {
+    super.setInput(input, seekForwardOnly, ignoreMetadata)
+    reset()
+  }
+
+  override fun getWidth(imageIndex: Int): Int {
+    loadInfoIfNeeded()
+    return width?.raw()?.toInt() ?: throw IOException("SVG document not loaded")
+  }
+
+  override fun getHeight(imageIndex: Int): Int {
+    loadInfoIfNeeded()
+    return height?.raw()?.toInt() ?: throw IOException("SVG document not loaded")
+  }
+
+  @Throws(IOException::class)
+  private fun loadInfoIfNeeded() {
+    val input = input
+    if (jSvgDocument == null && input is ImageInputStream) {
+
+      try {
+        // Not using ByteArray to avoid potential humongous allocation
+        ImageInputStreamAdapter(input).buffered().use {
+          jSvgDocument = createJSvgDocument(it).also { svg ->
+            width = svg.width
+            height = svg.height
+          }
+        }
+      } catch (e: IOException) {
+        // could not read the SVG document
+      }
```

Comment:
Need to catch `IllegalStateException` otherwise it may fail decoding the SVG, e.g. one from codecov, see https://github.com/weisJ/jsvg/issues/91

```text
Oct 03, 2024 11:35:03 AM com.github.weisj.jsvg.parser.css.impl.SimpleCssParser$a a
WARNING: Expected 'CURLY_OPEN' but got 'Token{type=CLASS_NAME, data='s'}'
Oct 03, 2024 11:35:03 AM com.github.weisj.jsvg.parser.css.impl.SimpleCssParser$a a
WARNING: Expected 'CURLY_OPEN' but got 'Token{type=EOF, data='null'}'
Exception in thread "main" java.lang.IllegalStateException: Cyclic dependency involving node 'white' detected.
	at com.github.weisj.jsvg.parser.ParsedElement.a(SourceFile:125)
	at com.github.weisj.jsvg.parser.ParsedElement.nodeEnsuringBuildStatus(SourceFile:82)
	at com.github.weisj.jsvg.parser.AttributeNode.b(SourceFile:1128)
	at com.github.weisj.jsvg.parser.AttributeNode.c(SourceFile:198)
	at com.github.weisj.jsvg.cS.a(SourceFile:61)
	at com.github.weisj.jsvg.ba.build(SourceFile:87)
	at com.github.weisj.jsvg.parser.ParsedElement.build(SourceFile:115)
	at com.github.weisj.jsvg.parser.ParsedElement.build(SourceFile:113)
	at com.github.weisj.jsvg.parser.ParsedElement.build(SourceFile:113)
	at com.github.weisj.jsvg.parser.ParsedElement.build(SourceFile:113)
	at com.intellij.ui.svg.JSvgDocumentFactoryKt.buildDocument(JSvgDocumentFactory.kt:110)
	at com.intellij.ui.svg.JSvgDocumentFactoryKt.createJSvgDocument(JSvgDocumentFactory.kt:37)
	at com.intellij.ui.svg.JSvgDocumentFactoryKt.createJSvgDocument$default(JSvgDocumentFactory.kt:34)
	at com.intellij.ui.svg.JSvgDocumentFactoryKt.createJSvgDocument(JSvgDocumentFactory.kt:26)
	at org.intellij.images.util.imageio.svg.SvgImageReader.loadInfoIfNeeded(SvgImageReader.kt:47)
	at org.intellij.images.util.imageio.svg.SvgImageReader.read(SvgImageReader.kt:60)
	at java.desktop/javax.imageio.ImageIO.read(ImageIO.java:1470)
	at java.desktop/javax.imageio.ImageIO.read(ImageIO.java:1422)
	at org.intellij.images.util.imageio.svg.QuickImageViewer.main(QuickImageViewer.kt:28)
```